File: Data Structures & Algorithms/subsets-ii/submission-11.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort() 
        res = [] 

        def backtracking(start, cur): 
            res.append(cur.copy()) 

            for i in range(start, len(nums)): 
                if i > start and nums[i] == nums[i - 1]: 
                    continue 

                cur.append(nums[i]) 
                backtracking(i + 1, cur) 
                cur.pop() 

        backtracking(0, []) 
        return res

File: Data Structures & Algorithms/three-integer-sum/submission-11.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        let res = [];
        nums = nums.sort((a,b) => a-b)

        for (let i = 0; i < nums.length; i ++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] === nums[i - 1]) continue;

            let l = i + 1;
            let r = nums.length - 1;

            while(l < r) {
                let cur = nums[i] + nums[l] + nums[r];
                if (cur < 0) {
                    l ++
                } else if (cur > 0) {
                    r --
                } else {
                    res.push([nums[i], nums[l], nums[r]]);
                    l ++;
                    while(l < r && nums[l] === nums[l - 1]) {
                        l ++
                    }
                }
            }
        }

        return res
    }
}
